lease.go

```diff
@@ -23,109 +23,119 @@ type WgRecord struct {
 	expires time.Time
 }
 
+func (wgr WgRecord) String() string {
+	return wgr.PubKey + " " + wgr.IP.String() + " " + wgr.expires.Format(time.RFC3339)
+}
+
 // FileLeaseManager implements functionality for managing address leases for
 // peers, using a file as a state backend.
 type FileLeaseManager struct {
-	wgRecords map[string]*WgRecord
-	filename  string
-	cidr      *net.IPNet
-	ip        net.IP
+	cidr           *net.IPNet
+	deviceName     string
+	filename       string
+	ip             net.IP
+	wgRecords      map[string]WgRecord
+	wgRecordsMutex sync.Mutex
 }
 
-func newFileLeaseManager(filename string, cidr *net.IPNet, ip net.IP) (*FileLeaseManager, error) {
-	if filename == "" {
+func newFileLeaseManager(cfg *serverConfig) (*FileLeaseManager, error) {
+	if cfg.LeasesFilename == "" {
 		return nil, fmt.Errorf("file name cannot be empty")
 	}
-	log.Printf("leases filename: %s\n", filename)
-	leaseDir := filepath.Dir(filename)
+	log.Printf("leases filename: %s\n", cfg.LeasesFilename)
+	leaseDir := filepath.Dir(cfg.LeasesFilename)
 	err := os.MkdirAll(leaseDir, 0755)
 	if err != nil {
 		log.Printf("Error: unable to create directory=%s", leaseDir)
 		return nil, err
 	}
 
-	wgRecords := make(map[string]*WgRecord)
-	r, err := os.Open(filename)
-	if err == nil {
-		defer r.Close()
-		wgRecords, err = loadWgRecords(r)
-		if err != nil {
-			return nil, err
-		}
-		log.Println("records loaded")
-	} else {
-		log.Printf("unable to open leases file: %v", err)
+	lm := &FileLeaseManager{
+		cidr:       cfg.WireguardIPNetwork,
+		deviceName: cfg.DeviceName,
+		filename:   cfg.LeasesFilename,
+		ip:         cfg.WireguardIPAddress,
 	}
 
-	lm := &FileLeaseManager{
-		wgRecords: wgRecords,
-		filename:  filename,
-		ip:        ip,
-		cidr:      cidr,
+	if err := lm.loadWgRecords(); err != nil {
+		return nil, err
 	}
 
-	if err := updateWgPeers(lm); err != nil {
+	if err := lm.updateWgPeers(); err != nil {
 		return nil, err
 	}
 
 	log.Println("Init complete")
 	return lm, nil
 }
 
-func loadWgRecords(r io.Reader) (map[string]*WgRecord, error) {
-	sc := bufio.NewScanner(r)
-	records := make(map[string]*WgRecord)
+func (lm *FileLeaseManager) loadWgRecords() error {
+	lm.wgRecordsMutex.Lock()
```

Comment:
Why do we need to lock it? Is there ever going to be more then one FileLeaseManager?

Reply:
Nevermind, I see, because we no longer append

Reply:
It's not that really but we have a map (`wgRecords`) that is being operated on from different goroutines: the main one, the ones from the http requests and the sync loop will all modify it. We really need to lock it before any read/write.